sdk/src/metrics/meter.cc

```diff
@@ -0,0 +1,775 @@
+#include "opentelemetry/sdk/metrics/meter.h"
+OPENTELEMETRY_BEGIN_NAMESPACE
+namespace sdk
+{
+namespace metrics
+{
+nostd::shared_ptr<metrics_api::Counter<short>> Meter::NewShortCounter(
+    nostd::string_view name,
+    nostd::string_view description,
+    nostd::string_view unit,
+    const bool enabled)
+{
+  if (!IsValidName(name) || NameAlreadyUsed(name))
```

Comment:
`NameAlreadyUsed` is not thread safe, correct?
For example, even if this check passed, there could still be race condition before we actually call `short_metrics_.insert`:
1. Thread A checked NameAlreadyUsed("foo") and it returned false.
2. Thread B checked NameAlreadyUsed("foo") and it returned false.
3. Thread A inserted the newly created "foo".
4. Thread B inserted the newly created "foo".

Reply:
Hmm that is a good point. Moving the mutex.locks() in the CreateInstrument functions to the top of the function definitions would resolve this issue, yes?

Reply:
Or, rather I think I may need to lock both the metrics_lock and observers_lock in the NameAlreadyUsed function. I believe the solution in my previous comment would still allow for a similar situation to occur if one instrument was synchronous and the other asynchronous.

Reply:
Ok, it required a tiny bit of refactoring but I believe that I have solved this issue now in an elegant way. Thanks for the feedback!